onnxruntime/core/framework/mem_pattern_planner.h

```diff
@@ -31,6 +31,89 @@ class MemPatternPlanner {
  public:
   MemPatternPlanner() = default;
 
+  void TraceAllocation(int ml_value_idx, std::vector<size_t> program_counter_start, std::vector<size_t> program_counter_end, size_t size) {
+    std::lock_guard<OrtMutex> lock(lock_);
+
+    if (size == 0) {
+      allocs_.emplace_back(ml_value_idx, MemoryBlock(0, 0));
+      return;
+    }
+
+    size_t current = 0;
+    size_t waste_bytes = std::numeric_limits<size_t>::max();
+    size_t best_offset = 0;
+    if (!blocks_.empty()) {
+      auto last_block = allocs_[*blocks_.rbegin()];
+      best_offset = last_block.block_.offset_ + last_block.block_.size_;
+    }
+
+    for (auto it = blocks_.begin(); it != blocks_.end(); it++) {
+            // Memory block can be re-used as long as there is no overlap between their time schedules.
+      bool overlap = false;
+      if (allocs_[*it].reuse_) {
+        //bool keep_looping = true;
+        //while(keep_looping);
+        ORT_ENFORCE(program_counter_start.size() == program_counter_end.size());
+        ORT_ENFORCE(program_counter_start.size() == program_counter_end.size());
+
+        // Ensure memory time schedule is sorted.
+        size_t start = 0;
+        for (size_t index = 0; index < program_counter_start.size(); index += 1) {
+          ORT_ENFORCE((program_counter_start[index] > start) || (start == 0));
+          ORT_ENFORCE(program_counter_start[index] <= program_counter_end[index]);
+          start = program_counter_start[index];
+        }
```

Comment:
Can this be done outside the loop?  As far as I can tell, it's an input and not being changed anywhere in this code.  
Better yet, can this check be done in `allocation_planner.cc` or where ever it's computed?

Reply:
Good idea.

---
In reply to: [505112197](https://github.com/microsoft/onnxruntime/pull/4443#discussion_r505112197) [](ancestors = 505112197)